From e2139df19b0cd3ba8b52c52184e4ca6df130da39 Mon Sep 17 00:00:00 2001
From: ezio84 <brabus84@gmail.com>
Date: Sat, 16 Jun 2018 16:57:19 +0200
Subject: [PATCH 22/48] GlobalScreenshot: Fix screenshot not saved with some
 languages

like Virgin Islands English when taking a screenshot of the Settings app

todo: find a less noob way, if possible, to handle this

Change-Id: Ic04f66f5813b9597c96835d15c74509c93530a5c
---
 .../systemui/screenshot/GlobalScreenshot.java        | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/packages/SystemUI/src/com/android/systemui/screenshot/GlobalScreenshot.java b/packages/SystemUI/src/com/android/systemui/screenshot/GlobalScreenshot.java
index a06251489a7..cd22ea3b942 100644
--- a/packages/SystemUI/src/com/android/systemui/screenshot/GlobalScreenshot.java
+++ b/packages/SystemUI/src/com/android/systemui/screenshot/GlobalScreenshot.java
@@ -84,6 +84,7 @@ import com.android.systemui.util.NotificationChannels;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
 import java.text.DateFormat;
 import java.text.SimpleDateFormat;
 import java.util.Date;
@@ -163,7 +164,16 @@ class SaveImageInBackgroundTask extends AsyncTask<Void, Void, Void> {
         boolean onKeyguard = context.getSystemService(KeyguardManager.class).isKeyguardLocked();
         if (!onKeyguard && appName != null) {
             // Replace all spaces and special chars with an underscore
-            String appNameString = appName.toString().replaceAll("[\\\\/:*?\"<>|\\s]+", "_");
+            String appNameString = appName.toString();
+            try {
+                // With some languages like Virgin Islands English, the Settings app gets a weird
+                // long name and some special voodoo chars, so we convert the string to utf-8 to get
+                // a  char instead, easy to remove it then
+                final String temp = new String(appNameString.getBytes("ISO-8859-15"), "UTF-8");
+                appNameString = temp.replaceAll("[]+", "");
+            } catch (UnsupportedEncodingException e) {}
+            // Now remove all special chars from the app name
+            appNameString = appNameString.replaceAll("[\\\\/:*?\"<>|\\s]+", "_");
             mImageFileName = String.format(SCREENSHOT_FILE_NAME_TEMPLATE_APPNAME,
                     appNameString, imageDate);
         } else {
-- 
2.17.1

